src/asignacion_aulica/backend/preferencias.py

```diff
@@ -43,65 +49,105 @@ def construir_edificios(aulas: DataFrame) -> dict[str, set[int]]:
 
-def obtener_cantidad_de_clases_fuera_del_edificio_preferido(clases: DataFrame, aulas: DataFrame, modelo: cp_model.CpModel, asignaciones: np.ndarray):
+def obtener_cantidad_de_clases_fuera_del_edificio_preferido(
+        clases: DataFrame,
+        aulas: DataFrame,
+        modelo: cp_model.CpModel,
+        asignaciones: np.ndarray
+    ) -> tuple[LinearExpr, int]:
     '''
-    Devuelve una expresión que representa la cantidad de clases fuera de su edificio preferido.
+    Devuelve una expresión que representa la cantidad de clases fuera de su
+    edificio preferido, y su cota superior.
     '''
     edificios = construir_edificios(aulas)
 
     # La cantidad comienza con el total de clases, y por cada clase que se
     # encuentra dentro de su edificio preferido se resta 1 a la cantidad de
     # clases fuera del edificio preferido
     cantidad_de_clases_fuera_del_edificio_preferido = len(clases)
+
     for clase in clases.itertuples():
         if clase.edificio_preferido:
             # Esta expresión da 1 o 0
             en_edificio_preferido = sum(asignaciones[clase.Index, aula] for aula in edificios[clase.edificio_preferido])
             cantidad_de_clases_fuera_del_edificio_preferido -= en_edificio_preferido
 
-    return cantidad_de_clases_fuera_del_edificio_preferido
+    cota_superior = len(clases)
+    return cantidad_de_clases_fuera_del_edificio_preferido, cota_superior
 
-def obtener_cantidad_de_alumnos_fuera_del_aula(clases: DataFrame, aulas: DataFrame, modelo: cp_model.CpModel, asignaciones: np.ndarray):
+def obtener_cantidad_de_alumnos_fuera_del_aula(
+        clases: DataFrame,
+        aulas: DataFrame,
+        modelo: cp_model.CpModel,
+        asignaciones: np.ndarray
+    ) -> tuple[LinearExpr, int]:
     '''
-    Devuelve una expresión que representa la cantidad de alumnos que exceden la 
-    capacidad del aula asignada a su clase.
+    Devuelve una expresión que representa la cantidad de alumnos que exceden la
+    capacidad del aula asignada a su clase, y su cota superior.
     '''
     cantidad_de_alumnos_fuera_del_aula = 0
+    cota_superior_total = 0
 
     for clase in clases.itertuples():
         exceso_de_capacidad = modelo.new_int_var(0, clase.cantidad_de_alumnos, f"exceso_de_capacidad_de_{clase.nombre}")
+        cota_superior = 0
+
         for aula in aulas.itertuples():
             asignada_a_este_aula = asignaciones[clase.Index, aula.Index]
 
-            if clase.cantidad_de_alumnos > aula.capacidad:
-                modelo.add(exceso_de_capacidad == clase.cantidad_de_alumnos - aula.capacidad).only_enforce_if(asignada_a_este_aula)
-            else:
-                modelo.add(exceso_de_capacidad == 0).only_enforce_if(asignada_a_este_aula)
-        
+            posible_exceso = max(0, clase.cantidad_de_alumnos - aula.capacidad)
+            modelo.add(exceso_de_capacidad == posible_exceso).only_enforce_if(asignada_a_este_aula)
+
+            if not isinstance(asignada_a_este_aula, np.integer):
```

Comment:
Las asignaciones constantes pueden ser int o np.integer, así que este chequeo no me genera mucha confianza.

```suggestion
            if isinstance(asignada_a_este_aula, cp_model.IntVar):
```

Reply:
Lo mismo en obtener_capacidad_sobrante